File: src/baekjoon/p15686/Main.java
package baekjoon.p15686;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class Main {
	static int N;
	static int[][] MAP;
	static List<int[]> chicken;

	static int dist(int y1, int x1, int y2, int x2) {
		return Math.abs(y1 - y2) + Math.abs(x1 - x2);
	}

	static int minDist(int y, int x) {
		int min = 100;
		for (int i = 0; i < N; i++) {
			for (int j = 0; j < N; j++) {
				if (MAP[i][j] == 2) {
					min = Math.min(min, dist(y, x, i, j));
				}
			}
		}
		return min;
	}

	static int cityDist(List<int[]> picked) {
		int sum = 0;
		for (int i = 0; i < N; i++) {
			for (int j = 0; j < N; j++) {
				if (MAP[i][j] == 1) {
					int min = 100;
					for (int[] chicken : picked) {
						int y = chicken[0]; int x = chicken[1];
						min = Math.min(min, dist(i, j, y, x));
					}
					sum += min;
				}
			}
		}
		return sum;
	}

	static int min = 250000;

	static void BFC(int i, int toPick, List<int[]> picked) {
		if (i == chicken.size()) {
			if (toPick == 0) {min = Math.min(min, cityDist(picked));}
			return;
		}
		BFC(i + 1, toPick, picked);
		picked.add(chicken.get(i));
		BFC(i + 1, toPick - 1, picked);
		picked.remove(picked.size() - 1);
 	}

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine(), " ");
		N = Integer.parseInt(st.nextToken());
		MAP = new int[N][N];
		chicken = new ArrayList<>();
		int M = Integer.parseInt(st.nextToken());
		for (int i = 0; i < N; i++) {
			st = new StringTokenizer(br.readLine(), " ");
			for (int j = 0; j < N; j++) {
				MAP[i][j] = Integer.parseInt(st.nextToken());
				if (MAP[i][j] == 2) {chicken.add(new int[] {i, j});}
			}
		}
		List<int[]> picked = new ArrayList<>();
		BFC(0, M, picked);
		System.out.println(min);
	}

}

File: src/baekjoon/p15686/Main.py
import sys
input = sys.stdin.readline
N, M = map(int, input().split())
chicken = []
open = []
home = []
for i in range(N):
    row = list(map(int, input().split()))
    for j in row :
        if row[j] == 1 : home.append((i, j))
        if row[j] == 2:
            chicken.append((i, j))
            open.append(True)

# p1과 p2사이의 거리를 반환
def dist(p1, p2) : return abs(p1[0] - p2[0]) + abs(p1[1] - p2[1])


# 좌표 here에서 가장 가까운 치킨 집 까지의 거리 반환
def minDist(here):
    ret = 10000
    for i in range(len(chicken)):
        print(open)
        if open[i] : ret = min(ret, dist(here, chicken[i]))
    return ret


# 도시의 치킨 거리 반환
def chickenDist():
    sum = 0
    for h in home : sum += minDist(h)
    return sum


# i번째 치킨집 부터 k개를 폐업하는 경우 도시의 최소 치킨 거리
def btk(i, k):
    if i == len(chicken) : return chickenDist() if k == 0 else 987654321
    ret = btk(i + 1, k) # i번째 치킨집을 폐업하지 않는 경우
    if k :
        open[i] = False
        ret = min(ret, btk(i + 1, k - 1))
        open[i] = True
    return ret

print(chicken)
print(btk(0, M))
